Wait for interface to appear before querying for it.

amshinde opened the pull request:
We were querying for mac address and interface name before
the wait_for_nic call which waits for interface name change
uevent. Also wait on the event only when the interface name
is provided.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

jodh-intel commented:
This looks good, but this sort of change really needs a comment in the code to explain the ordering of these calls.

dlespiau commented on src/net.c:
```diff
@@ -991,6 +991,13 @@ static int hyper_setup_interface(struct rtnl_handle *rth,
 	req.n.nlmsg_type = RTM_NEWADDR;
 	req.ifa.ifa_family = AF_INET;
 
+	if (iface->device) {
```
So now we have the following checks on iface->device:

````
if (iface->device) {
     wait_for_nic(iface->device)
         if (!iface->device)
             return -1
````

It seems we could write nicer checks in general. It feels a bit like we end up checking the same pointer at every function call. In this case, maybe adding saying that wait_for_nic() accepts a NULL as a valid value (saying: iface->device is optional, waiting on NULL device name is just a no-op) could make the code a bit nicer I think.

This is similar to how go treats nil is some cases. nil is a valid input for `len()`, it just returns 0 instead of an error and that allows removing some error paths.

Just a general comment, you don't need to address it.

amshinde replied:
@dlespiau I agree, we do tend to check the same pointers at every call. I like your suggestion, I'll push a change for this.

amshinde committed:
Treat null interface as NOP in wait_for_nic function

Instead of returning error, treat a NULL interface as
valid input and return 0 instead. This is to minimise error
checking paths.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde committed:
Wait for interface to appear before querying for it

We were querying for mac address and interface name before
the wait_for_nic call which waits for interface name change
uevent.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde commented:
@jodh-intel @dlespiau Changes pushed

dlespiau commented on src/net.c:
```diff
@@ -991,6 +990,18 @@ static int hyper_setup_interface(struct rtnl_handle *rth,
 	req.ifa.ifa_family = AF_INET;
 
+	/* Check and wait for the interface to appear in /sys first if the 
+	 * predictable  name is provided. A process like systemd/udev may be
```
two spaces between "predictable" and "name"!

dlespiau replied:
might as well give the real process name `systemd-udevd`

amshinde replied:
fixed

dlespiau commented:
lgtm

sboeuf commented:
LGTM
